plutus-contract/src/Wallet/Emulator/Chain.hs

```diff
@@ -196,31 +196,39 @@ mkValidationEvent idx t result events =
 -- | Validate a transaction in the current emulator state.
 validateEm
     :: S.MonadState Index.ValidationCtx m
-    => Params
-    -> Slot
+    => Slot
     -> Validation.UTxO Index.EmulatorEra
     -> CardanoTx
     -> m (Maybe Index.ValidationErrorInPhase, [ScriptValidationEvent])
-validateEm params h cUtxoIndex txn = do
-    ctx@(Index.ValidationCtx idx _) <- S.get
-    let ((e, idx'), events) = txn & mergeCardanoTxWith
+validateEm h cUtxoIndex txn = do
+    ctx@(Index.ValidationCtx idx params) <- S.get
+    let (e, events) = txn & mergeCardanoTxWith
             (\tx -> Index.runValidation (Index.validateTransaction h tx) ctx)
-            (\tx -> ((validateL params h cUtxoIndex tx, idx), []))
-            (\((e1, utxo), sve1) ((e2, _), sve2) -> ((e1 <|> e2, utxo), sve1 ++ sve2))
-    _ <- S.put ctx{Index.vctxIndex=idx'}
+            (\tx -> validateL params h cUtxoIndex tx)
+            (\(e1, sve1) (e2, sve2) -> (e1 <|> e2, sve1 ++ sve2))
+        idx' = case e of
+            Just (Index.Phase1, _) -> idx
+            Just (Index.Phase2, _) -> Index.insertCollateral txn idx
+            Nothing                -> Index.insert txn idx
+    _ <- S.put ctx{ Index.vctxIndex = idx' }
     pure (e, events)
 
-validateL :: Params -> Slot -> Validation.UTxO Index.EmulatorEra -> SomeCardanoApiTx -> Maybe Index.ValidationErrorInPhase
-validateL params slot idx (CardanoApiEmulatorEraTx tx) = Validation.hasValidationErrors params (fromIntegral slot) idx tx
+-- TODO: generate ScriptValidationEvents
```

Comment:
What do that refer to exactly? Do we a story for that?

Reply:
Good point, I will make one.